modules/build/src/main/scala/scala/build/internal/MainClass.scala

```diff
@@ -37,26 +43,47 @@ object MainClass {
       if (foundMainClass) nameOpt else None
   }
 
-  def findInClass(path: os.Path): Iterator[String] = {
-    val is = os.read.inputStream(path)
+  def findInClass(path: os.Path): Iterator[String] =
+    findInClass(os.read.inputStream(path))
+  def findInClass(is: InputStream): Iterator[String] =
     try {
       val reader  = new ClassReader(is)
       val checker = new MainMethodChecker
       reader.accept(checker, 0)
       checker.mainClassOpt.iterator
     }
     finally is.close()
+
+  def findInJar(path: os.Path): Iterator[String] = {
+    val content        = os.read.bytes(path)
+    val jarInputStream = WrappedZipInputStream.create(new ByteArrayInputStream(content))
+    val classesInputStreams: Seq[InputStream] =
+      jarInputStream.entries().foldLeft(List.empty[InputStream]) {
```

Comment:
NIT Why do we need a `foldLeft` here? Wouldn't a `flatMap` do?

Reply:
Plus we could keep it as an `Iterator` right from the start it seems.

Reply:
ah, good catch.
should be fixed now.